// iosync/index.js
module.exports = function (opt) {
  def_opt = {};
  if(!opt.server){
    def_opt.server = require('http').createServer();
  }

  var io = require('socket.io')(opt.server || def_opt.server, opt.socket_io);

  if(opt.session){
    var sharedsession = require("express-socket.io-session");
    io.use(sharedsession(opt.session));
  }

  io.on("connection", function(socket) {

    socket.on("login", function(login_data, client_callback) {
      if(callbacks.login){
        var response = callbacks.login(login_data, socket.handshake.session, function (response) {
          if (!response.error){
            socket.handshake.session.user = {
              id: response.user.id,
              capabilities: response.user.capabilities
            };
            socket.handshake.session.save();
            client_callback(response.client_params);
          }else client_callback(response);
        });
      } else client_callback({error: "login_system_is_not_configured"});
    });

    socket.on("logout", function(data, client_callback) {
      if(callbacks.logout){
        var response = callbacks.logout(socket.handshake.session, function (response) {
          if (!response.error){
            delete socket.handshake.session.user;
            socket.handshake.session.save();
            client_callback(response.client_params);
          } else client_callback(response);
        });
      } else client_callback({error: "login_system_is_not_configured"});
    });

    socket.on("check", function (option, client_callback) {
      switch (option) {
        case 'login':
          client_callback(!!socket.handshake.session.user);
          break;
      }
    })

    socket.on('bind', function (path) {

    });
    socket.on('unbind', function (path) {

    });

    socket.on('patch', function (patch) {
      console.log(patch);
    });

    socket.on("disconnect", function() {
    });

  });

  if(def_opt.server) def_opt.server.listen(opt.port || process.env.PORT || 80);

  var callbacks = {
    login: undefined,
    logout: undefined
  }

  return {
    on: function (event, callback) {
      switch (event) {
        case "login":
          callbacks.login = callback;
          break;
        case "logout":
          callbacks.logout = callback;
          break;
        default:
        throw new unknown_event();
      }
    },
    off: function (event) {

    }
  }

  function unknown_event(){
    this.message = "Unknown event";
    this.toString = function () {
      return this.message;
    };
  }
}
